src/clj/rems/application/search.clj

```diff
@@ -93,24 +92,61 @@
         (doseq [[k v] terms]
           (.add doc (TextField. (name k) v Field$Store/NO)))
         (.add doc (TextField. "all" (str/join " " (vals (into (sorted-map) terms))) Field$Store/NO))
-        (.updateDocument writer (Term. app-id-field app-id) doc))
+        (.updateDocument writer (Term. app-id-field app-id) doc)
+        (log/info "Indexed application" app-id))
       (catch Throwable t
         (throw (Error. (str "Error indexing application " app-id) t))))))
 
-(defn refresh! []
+(defn index-events! [events get-full-internal-application]
   (locking index-lock
-    (let [{::keys [directory ^SearcherManager searcher-manager last-processed-event-id]} @search-index
-          events (events/get-all-events-since last-processed-event-id)]
+    (let [{::keys [directory ^SearcherManager searcher-manager]} @search-index]
       (when-not (empty? events)
         (with-open [writer (IndexWriter. directory (-> (IndexWriterConfig. analyzer)
                                                        (.setOpenMode IndexWriterConfig$OpenMode/APPEND)))]
           (let [app-ids (distinct (map :application/id events))]
-            (log/info "Start indexing" (count app-ids) "applications...")
-            (doseq [app-id app-ids]
-              (index-application! writer (applications/get-application app-id)))
-            (log/info "Finished indexing" (count app-ids) "applications")))
-        (.maybeRefresh searcher-manager)
-        (swap! search-index assoc ::last-processed-event-id (:event/id (last events)))))))
+            #_(log/info "Start indexing" (count app-ids) "applications...")
+            (doseq [app-id app-ids
+                    :let [application (get-full-internal-application app-id)]
+                    :when application]
+              (index-application! writer application))
+            #_(log/info "Finished indexing" (count app-ids) "applications")))
+        (.maybeRefresh searcher-manager)))))
+
+(def indexing-requests (atom nil))
+
+(defn process-index-events! []
+  (locking index-lock
+    (when-some [requests (seq @indexing-requests)]
+      (let [events (mapcat first requests)
+            get-full-internal-application (second (first requests))] ; NB: doesn't matter which get-full-internal-application is actually used
+        (index-events! events
+                       get-full-internal-application))  ; TODO avoid by moving to system ns
+      (reset! indexing-requests nil))))
```

Comment:
- Locking not necessary with `reset-vals!`
- `get-full-internal-application` could be static reference if we move the system ns